### PR TITLE
Documentation

### PR DESCRIPTION
**Pull request by U8NWXD**
**Deadline:** September 18th
* Remove outdated material and add disclaimers for missing sections

---------

Docs are currently being built on readthedocs: https://vivarium-core.readthedocs.io/en/docs/. This is the version of the docs from the `docs` branch, but `latest` and `stable` will update once we merge this PR.

Summary:

* Add the hand-written documentation from the old Vivarium repo
* Auto-generate docs from docstrings
* Move example glucose phosphorylation process, compartment, and experiment into vivarium-core from vivarium-cell since they're used in the docs for demonstration
* Move injector from vivarium-cell into vivarium-core since it is generally useful and used in the docs
* Fix references and syntax throughout the docstrings and written docs
* Update release script to clear `dist` directory

TODO:

* [x] Re-write docs to reflect recent changes
* [x] Add tests back and run in CI
* [ ] Give @eagmon and @prismofeverything access on readthedocs
* [x] Remove the code moved in from vivarium-cell from vivarium-cell
* [x] Make sure readthedocs webhook is working
* [x] Update getting started guide to remove zookeeper and kafka (and rest of docs)
* [x] Update docs link in README
* [ ] Update description of modeling paradigm
* [x] Write docs for package installation
* [x] Create template repo

### CODE REVIEW

**Comment by eagmon** [comment]
@U8NWXD -- thanks for getting this migrated and updated!

**Comment by eagmon** [comment]
@U8NWXD -- do you have any thoughts about how to combine process references from other libraries into this documentation?  For example, to https://vivarium-core.readthedocs.io/en/docs/reference/api/vivarium.processes.html.  ```vivarium-cell``` has many processes with good documentation.

**Comment by U8NWXD** [comment]
@eagmon I was thinking of having two documentation sites: one for vivarium-core and one for vivarium-cell. Vivarium-cell would have the documentation on the individual processes.

I'm trying to avoid referencing vivarium-cell processes in vivarium-core to keep the vivarium-core docs self-contained. I'll have links in vivarium-cell to vivarium-core, but not the other way around, to avoid circular references. The idea here is that vivarium-core should not depend in any way on how other people use it, e.g. in vivarium-cell, including documentation. In this PR I created some basic processes, compartments, and experiments in vivarium-core so we can demonstrate how vivarium-core can be used without making people look at vivarium-cell. Does this make sense?

**Comment by U8NWXD** [comment]
@eagmon This PR is almost ready. I've removed the old references to zookeeper and Kafka, and I updated the manually written docs to reflect our recent changes. One thing left is that the [description of Vivarium's modeling paradigm](https://vivarium-core.readthedocs.io/en/docs/guides/modeling_paradigm.html) still focuses on agent-based and whole-cell modeling. We should think about whether this should change

**Comment by U8NWXD** [comment]
Another note: The docs are written around people cloning the GitHub repo still instead of installing the pip package. This is helpful because the repo has a demo process, compartment, and experiment users can experiment with, but we don't actually want people to use the code this way. We could create a separate demo repository for users to get started with? This could be that template repo we discussed at one point.

File: AUTHORS.md
Comment (eagmon) [inline AUTHORS.md:11]: jerry.morrison

File: doc/_static/draw_tree.py
Comment (eagmon) [inline doc/_static/draw_tree.py:31]: Looks intriguing! I didn't notice this before. Is this displayed in the docs?

File: doc/getting_started.rst
Comment (eagmon) [inline doc/getting_started.rst:142]: I believe this is only needed for ```vivarium-cell```... Points to the issue of having multiple possible ```getting_started``` docs for different repos.

File: doc/guides/compartments.rst
Comment (eagmon) [inline doc/guides/compartments.rst:203]: we now link to stores in tuples to specify their relative compartment: ```('cell',)```. Might be worth giving an example of going up one level: ```('..', 'boundary')```

Edit -- you describe this below already. That looks sufficient.

File: doc/guides/compartments.rst
Comment (eagmon) [inline doc/guides/compartments.rst:273]: typo - "hierarchy guide"

File: doc/guides/hierarchy.rst
Comment (eagmon) [inline doc/guides/hierarchy.rst:141]: typo -- "Hierarchy"

**Review by eagmon — APPROVED** [review]
Thanks @U8NWXD -- this looks great. Thank you for noting the needed update on ```modeling_paradigm```, I can take care of that in the next week or two.

I definitely think we should make a template repo. How's the name ```vivarium-template```? How can we coordinate the docs for that template and the core engine? Do you know of any other project that has the multiple library coordination challenge as we do?

Just a couple requests before you merge: Can you build this, add a link to the main core ```README```, and add me and Ryan to readthedocs.  Do we also have web hooks in place to automatically build the docs, like we had in the old vivarium?

File: AUTHORS.md
Comment (U8NWXD) [reply AUTHORS.md:11]: This is actually how Jerry asked that his email be specified. Gmail ignores dots when resolving email addresses, so he'll still get emails to this address. However, if a spammer gets this address, Jerry can tell Gmail to delete emails to `j.erry.morrison@...` to block the spam. As a side note, Gmail also ignores everything after `+` but before `@` so you can do things like this: `example+<anything you like>@gmail.com`

File: doc/_static/draw_tree.py
Comment (U8NWXD) [reply doc/_static/draw_tree.py:31]: Yeah, it's in the [hierarchy guide](https://vivarium-core.readthedocs.io/en/docs/guides/hierarchy.html). The figure needs some aesthetic work though

**Comment by U8NWXD** [comment]
Regarding template repo docs: My suggestion would be to not have docs for the template repo. That README should point to the vivarium-core docs. The vivarium-core docs can then link to the files in the template repo on GitHub instead of linking to their auto-generated docs. I don't see much point in having auto-generated docs for the template repo since the demo files aren't meant to be used in any models.

File: doc/guides/compartments.rst
Comment (U8NWXD) [reply doc/guides/compartments.rst:203]: I'll fix this part to use paths to be consistent. I also describe paths in more detail in the hierarchy guide

File: doc/guides/compartments.rst
Comment (U8NWXD) [reply doc/guides/compartments.rst]: Fixed in 7605b0d

File: doc/guides/hierarchy.rst
Comment (U8NWXD) [reply doc/guides/hierarchy.rst]: Fixed in 7605b0d

File: doc/getting_started.rst
Comment (U8NWXD) [reply doc/getting_started.rst]: removed in f1f742fff55012cf7398b512f3ef1928fbdfbc14

**Comment by U8NWXD** [comment]
> Just a couple requests before you merge: Can you build this, add a link to the main core README, and add me and Ryan to readthedocs. Do we also have web hooks in place to automatically build the docs, like we had in the old vivarium?

Read the Docs is complaining that the webhook isn't working, but the documentation is building automatically like it should, so I'm ignoring the warning. For example, you can see that the typo in the hierarchy guide is fixed: [link](https://vivarium-core.readthedocs.io/en/docs/guides/hierarchy.html#hierarchy-paths).

This PR also changes the README to point to the updated docs. Once we merge the PR, Read the Docs will build the documentation for `master` and the link will point to those built docs.

I've added you as a maintainer on Read the Docs. @prismofeverything could you send me your username for Read the Docs so I can add you too?

File: setup.py
Comment (U8NWXD) [inline setup.py:27]: @eagmon does this description look good to you? If so, we can make the GitHub repo description match for consistency

**Comment by U8NWXD** [comment]
@prismofeverything @eagmon FYI I added the tests I run on the documentation to GitHub actions now, so any changes that break the docs will be flagged on PRs. There are 2 tests I'm running:

1. A linter to check for typos and ensure the documentation writes phrases with multiple spellings consistently (e.g. timeseries vs time series vs time-series)
2. A test that builds the documentation and makes sure there are no errors. This ensures we don't have broken links, for example.

File: setup.py
Comment (eagmon) [reply setup.py:27]: How about this slightly modified description: `Engine for composing and simulating computational biology models with the Vivarium interface`. Yes to consistency with the repo description!

Comment (U8NWXD) [reply setup.py]: Done in 0bdd01e. I've also updated the repo description

**Comment by U8NWXD** [comment]
The template repository is up at https://github.com/vivarium-collective/vivarium-template, and I've re-written the getting started guide to use it

**Comment by U8NWXD** [comment]
@eagmon I think this is ready to go in now. I'll wait until later today to merge though in case you notice something I need to fix

**Comment by eagmon** [comment]
@U8NWXD -- this looks excellent, thank you! I'm excited about that template repo, we might already have a chance to use it today. Can you point me to those tests you mentioned?

**Comment by U8NWXD** [comment]
@eagmon the template repo won't work quite yet. It needs vivarium-core to have some processes that are introduced by this PR

**Comment by U8NWXD** [comment]
You can see the tests running here: https://github.com/vivarium-collective/vivarium-core/pull/6/checks?check_run_id=1129048015. They're defined here: https://github.com/vivarium-collective/vivarium-core/blob/docs/.github/workflows/docs.yml

**Pull request merged**
